tst/testinstall/MatrixObj/testmatobj_new.g

```diff
@@ -0,0 +1,299 @@
+            for dim in dims do 
+                for dom in doms do 
+                    if dim[1] = dim[2] then 
+                        Add(examples, [RandomInvertibleMat(opt.randomSource, dim[1], dom), dim[1], dim[2], dom]);
+                    fi;
+                od;
+            od;
+        else 
+            for dim in dims do 
+                for dom in doms do 
+                    Add(examples, [RandomMat(opt.randomSource, dim[1], dim[2], dom), dim[1], dim[2], dom]);
+                od;
+            od;
+        fi;
+    fi;
+
+    if IsBound(opt.sourceOfTruth) then 
+        return List(examples, x-> 
+            rec(matObj := NewMatrix(filter, x[4], x[3], x[1]),
+            filter := filter,
+            baseDomain := x[4],
+            nrCols := x[3],
+            nrRows := x[2],
+            mat := x[1],
+            sourceOfTruth := NewMatrix(opt.sourceOfTruth, x[4], x[3], x[1]))
+            );
+        #return List(examples, x -> [NewMatrix(filter, x[3], x[2], x[1]), NewMatrix(opt.sourceOfTruth, x[3], x[2], x[1])]);
+    else 
+        return List(examples, x-> 
+            rec(matObj := NewMatrix(filter, x[4], x[3], x[1]),
+            filter := filter,
+            baseDomain := x[4],
+            nrCols := x[3],
+            nrRows := x[2],
+            mat := x[1],
+            sourceOfTruth := x[1])
+        );
+        #return List(examples, x -> [NewMatrix(filter, x[3], x[2], x[1]), x[1]]);
+    fi;
+end;
+
+MatObjTest_TestBaseDomain := function(ex, opt, errors)
+    local domain;
+
+    domain := MatObjTest_CallFunc(BaseDomain,[ex.matObj], opt.breakOnError, errors, ex);
+
+    if domain <> fail then
+        if domain <> ex.baseDomain then
+            MatObjTest_HandleErrorWrongResult("BaseDomain", [ex.matObj], opt.breakOnError, ex, errors);
+        fi;
+    fi;
+end;
+
+MatObjTest_TestNrRows := function(ex, opt, errors)
+    local nrRows;
+
+    nrRows := MatObjTest_CallFunc(NrRows, [ex.matObj], opt.breakOnError, errors, ex);
+
+    if nrRows <> fail then
+        if nrRows <> ex.nrRows then
+            MatObjTest_HandleErrorWrongResult("NrRows", [ex.matObj], opt.breakOnError, ex, errors);
+        fi;
+    fi; 
+end;
+
+MatObjTest_TestNrCols := function(ex, opt, errors)
+    local nrCols;
+
+    nrCols := MatObjTest_CallFunc(NrCols, [ex.matObj], opt.breakOnError, errors, ex);
+    
+    if nrCols <> fail then
+        if nrCols <> ex.nrCols then
+            MatObjTest_HandleErrorWrongResult("NrCols", [ex.matObj], opt.breakOnError, ex, errors);
+        fi;
+    fi; 
+end;
+
+MatObjTest_TestMatElm := function(ex, opt, errors)
+    local col, row, elm;
+
+    col := ex.nrCols;
+    row := ex.nrRows;
+    elm := MatObjTest_CallFunc(MatElm, [ex.matObj, row, col], opt.breakOnError, errors, ex);
+
+    if elm <> fail then
+        if elm <> ex.sourceOfTruth[row,col] then
+            MatObjTest_HandleErrorWrongResult("MatElm", [ex.matObj, row, col], opt.breakOnError, ex, errors);
+        fi;
+    fi;
+end;
```

Comment:
As discussed: perhaps `MatObjTest_CallFunc` and  `MatObjTest_HandleErrorWrongResult` could be called by the code invoking `MatObjTest_TestMatElm` etc., and the code here would become something like this:
```suggestion
MatObjTest_TestMatElm := function(ex)
    local col, row, elm;
    for row in [1..NrRows(ex.matobj)] do
        for col in [1..NrCols(ex.matobj)] do
            elm := ex.matobj[row, col];
            if elm <> ex.sourceOfTruth[row,col] then
                return false;
            fi; 
        od;
    od;
    return true;
```
We'd loose some level of information in the error report, but we'd make it a lot easier to write and debug these test cases, which is not irrelevant given that we need to write a lot, and then people implementing matrixObj need to interact with them.

My argument in the discussion we just had was that the lost information is not important: to debug this, as a developer I would turn on `breakOnError` and re-run the test to debug it and find out the position the error occurred. Then I might want to write a stand-alone regression test case; for this it is of course very helpful that you print a `NewMatrix(...)` call, but that's enough for me. (If we really want more info, we can always add back a way to do that: e.g. return a string instead of a bool, or a pair `[result_bool, error_desc]`; or provide a callback function `MatObjTest_AddErrorDetails` (or whatever). But for now I'd just leave this out, and wait what our experience is with using the test suite.